pkg/controller/certificates/util.go

```diff
@@ -264,13 +264,13 @@ func GenerateLocallySignedTemporaryCertificate(crt *cmapi.Certificate, pkData []
 
 // RenewBeforeExpiryDuration will return the amount of time before the given
 // NotAfter time that the certificate should be renewed.
-func RenewBeforeExpiryDuration(notBefore, notAfter time.Time, specRenewBefore *metav1.Duration) time.Duration {
-	renewBefore := cmapi.DefaultRenewBefore
+func RenewBeforeExpiryDuration(notBefore, notAfter time.Time, specRenewBefore *metav1.Duration, defaultRenewBeforeExpiryDuration time.Duration) time.Duration {
+	renewBefore := defaultRenewBeforeExpiryDuration
 	if specRenewBefore != nil {
 		renewBefore = specRenewBefore.Duration
 	}
 	actualDuration := notAfter.Sub(notBefore)
-	if renewBefore > actualDuration {
+	if renewBefore >= actualDuration {
```

Comment:
This prevents the specific issue reported in #3396 where the vault issued certificate duration (30d) exactly matched the default renewBefore value (30d) causing every certificate to be immediately renewed.

But this still leaves (rare) edge cases :
*  CA might for some reason issue a certificate of 30d1s 
*  administrator sets defaultRenewBefore to 1d which would clash with 1d duration certificate
* Developer creates a Certificate with renewBefore 7d and duration 3d, but CA signs certificate for 7d
* ...or even the case where CA returns a certificate with a NotAfter set in the past
* or where NotAfter is set lower than NotBefore 


In all these cases there would be weird behaviour and potentially a large number of certificate requests generated.

There should be sanity checks on the NotBefore / NotAfter dates, tests to ensure that we don't ever calculate a negative renewal date  or  a renewal date in the past, and maybe some sort of backoff  to limit the rate of new certificate requests for an individual certificate.

Reply:
Nice fix here, handling all these edge cases would be a good issue to write up :)

Reply:
See https://github.com/jetstack/cert-manager/issues/3486